File: Interview Cake/7. Temperature Tracker/tempTrakcer.py
class TempTracker:
	def __init__(self):
		self.data = {}
		self.min = None
		self.max = None
		self.modeCount = None
		self.modeVal = None
		self.count = 0
		self.total = 0
	
	def insert(temp):
		if not data:
			self.min = self.max = self.modeVal = self.total = temp
			self.count = self.modeCount = 1

		self.data[temp] = self.data.get(temp, 0) + 1
		self.min = min(self.min, temp)
		self.max = max(self.max, temp)
		if data[temp] > self.modeCount:
			self.modeCount = data[temp]
			self.modeVal = temp
		self.count += 1
		self.total += temp

	def get_max():
		return self.max

	def get_min():
		return self.min

	def get_mean():
		return total/float(count)

	def get_mode():
		return self.modeVal

File: Interview Cake/7. Temperature Tracker/tempTrakcer.cpp
class TempTracker {
	private:

	int min_temp, max_temp, count, total, mode, modeCount;
	unordered_map<int, int> data;

	public:

	TempTracker() : min_temp(numeric_limits<int>::max()), max_temp(numeric_limits<int>::min()), 
	count(0), total(0), mode(0), modeCount(0) {}

	void insert(int temp) {
		min_temp = min(min_temp, temp);
		max_temp = max(max_temp, temp);
		count++;
		total += temp;
		if (data.find(temp) == data.end()) {
			data[temp] = 1;
		} else {
			data[temp] += 1;
		}

		if (data[temp] > modeCount) {
			modeCount = data[temp];
			mode = temp;
		}
	}

	int getMax() {
		return max_temp;
	}

	int getMin() {
		return min_temp;
	}

	double getMean() {
		return total/(double) count;
	}

	int getMode() {
		return mode;
	}
}